### PR TITLE
[CARBONDATA-3512]Index Server Enhancement

### PR DESCRIPTION
**Pull request by BJangir**
### What changes are proposed 

1. Remove the keytab dependency for IndexServer 
  	Currently IndexServer needs to configure keytab and prinicipal for both Client side and Server Side.But indexServer is super user and having super user's keytab and principal in client is not correct(specialy spark-submit).
  	Since IndexServer is wrapped around spark application so no need to ask Keytab from User for IndexServer. 
  2. Authentication :- This happens in org.apache.hadoop.security.SaslRpcClient#createSaslClient .it checks  getServerPrincipal(spark.carbon.indexserver.principal) and Server protocol (UGI of IndexServer). User need to configure spark.carbon.indexserver.principal properly.

  3. Authorization(ACL):- Support User who can access the IndexServer. 
  	Authorization is controlled by hadoop.security.authorization parameter.  	 
  	IndexServer has below scenarios.
  	1. Spark-submit,spark-shell,spark-Sql :-> These type of spark Application has UGI where LoginUser and LoginUser will be same either based on kinit or based on spark.yarn.principal. 
  		Authorization is done in org.apache.hadoop.ipc.Server#authorize using IndexServer ProtocolClass and  ACL list which is prepared  by org.apache.hadoop.security.authorize.PolicyProvider(generally hadoop-policy.xml with key security.indexserver.protocol.acl , by default *). 
  	2. Spark JDBCServer :- It has UGI based on ProxyUser like  user1(auth:PROXY)via spark/<DOMAIN>/<REALM>. where user1 is currentUser and spark is LoginUser (JDBCServer started UGI).
  		This type of  Authorization happens in org.apache.hadoop.security.authorize.ProxyUsers#authorize with proxyUserAcl list prepared by  hadoop.proxyuser.<INDEXSERVER_UGI>.users ,hadoop.proxyuser.<INDEXSERVER_UGI>.hosts ,hadoop.proxyuser.<INDEXSERVER_UGI>.groups. 
  		check IndexServer document for detail about configuration.

4. TokenRenewer:- IndexServer is NOT Token based Hadoop Service. It does not required Delegation Token as IndexServer does not connect to KDC since it is inside SparkApplication(both, Indexclient and IndexServer) so take advantage of it. 

image

Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/412/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/412/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/412/

File: docs/index-server.md [hunk 1]
```diff
@@ -180,6 +179,12 @@ that will authenticate the user to access the index server and no other service.
 | Name     |      Default Value    |  Description |
 |:----------:|:-------------:|:------:       |
 | ipc.client.rpc-timeout.ms |  NA | Set the above property to some appropriate value based on your estimated query time. The best option is to set this to the same value as spark.network.timeout. |
+| hadoop.security.authorization |  false | property to enable the Hadoop Security Enable.(Need only Server Side). |
```
Comment (akashrn5) [inline docs/index-server.md]: Change it from `property to enable the Hadoop Security Enable.(Need only Server Side). `  to `Property to enable the hadoop security which is required only on the server side.`

File: docs/index-server.md [hunk 2]
```diff
@@ -180,6 +179,12 @@ that will authenticate the user to access the index server and no other service.
...
+| hadoop.proxyuser.<indexserver_user>.users |  NA | User list for which IndexServer permission to be given ,check https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Superusers.html|
```
Comment (akashrn5) [inline docs/index-server.md]: I think `List of users for whom, IndexServer permissions were to be given` sounds better. Please cross check

File: docs/index-server.md [hunk 3]
```diff
@@ -180,6 +179,12 @@ that will authenticate the user to access the index server and no other service.
...
+| hadoop.proxyuser.<indexserver_user>.hosts |  NA | hosts list for which IndexServer permission to be given ,check https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Superusers.html|
```
Comment (akashrn5) [inline docs/index-server.md]: Same as above.

File: docs/index-server.md [hunk 4]
```diff
@@ -180,6 +179,12 @@ that will authenticate the user to access the index server and no other service.
...
+| hadoop.proxyuser.<indexserver_user>.groups |  NA | groups list for which IndexServer permission to be given ,check https://hadoop.apache.org/docs/current/hadoop-project-dist/hadoop-common/Superusers.html|
+| security.indexserver.protocol.acl |  * | List of User to be Authorized for Other than proxy Spark Application |
```
Comment (akashrn5) [inline docs/index-server.md]: same as above

File: integration/spark-common/src/main/scala/org/apache/carbondata/events/Events.scala [hunk 5]
```diff
@@ -198,3 +198,10 @@ trait BuildDataMapEventsInfo {
   val identifier: AbsoluteTableIdentifier
   val dataMapNames: scala.collection.mutable.Seq[String]
 }
+
+/**
+ * EventInfo for IndexServer
+ */
+trait IndexServerEventInfo {
+  val absTable: AbsoluteTableIdentifier
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/carbondata/events/Events.scala]: change `absTable` to `absoluteTableIdentifier`

File: integration/spark-common/src/main/scala/org/apache/carbondata/events/IndexServerEvents.scala [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.carbondata.events
+
+import org.apache.spark.sql.SparkSession
+
+import org.apache.carbondata.core.metadata.AbsoluteTableIdentifier
+
+case class IndexServerEvent(sparkSession: SparkSession, absTable: AbsoluteTableIdentifier)
```
Comment (akashrn5) [inline integration/spark-common/src/main/scala/org/apache/carbondata/events/IndexServerEvents.scala]: same as above

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 7]
```diff
@@ -105,6 +109,11 @@ object IndexServer extends ServerInterface {
         sparkSession.sparkContext.setLocalProperty("spark.jobGroup.id", request.getTaskGroupId)
         sparkSession.sparkContext
           .setLocalProperty("spark.job.description", request.getTaskGroupDesc)
+        // Fire Generic Event like ACLCheck..etc
```
Comment (akashrn5) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: if the event is getting fired only in get splits, one place, then better to make the operationContext as local variable

File: docs/index-server.md (hunk 1 above)
Comment (BJangir) [reply docs/index-server.md]: OK

File: docs/index-server.md (hunk 2 above)
Comment (BJangir) [reply docs/index-server.md]: OK

File: integration/spark-common/src/main/scala/org/apache/carbondata/events/Events.scala (hunk 5 above)
Comment (BJangir) [reply integration/spark-common/src/main/scala/org/apache/carbondata/events/Events.scala]: Done

File: integration/spark-common/src/main/scala/org/apache/carbondata/events/IndexServerEvents.scala (hunk 6 above)
Comment (BJangir) [reply integration/spark-common/src/main/scala/org/apache/carbondata/events/IndexServerEvents.scala]: Done

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala (hunk 7 above)
Comment (BJangir) [reply integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: operationContext  will used in all method except ShowCache() Method and those method will be called multiple time so defined global one time object should be ok.

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/453/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/454/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/455/

File: integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala [hunk 8]
```diff
@@ -83,7 +87,7 @@ object IndexServer extends ServerInterface {
   private val isExecutorLRUConfigured: Boolean =
     CarbonProperties.getInstance
       .getProperty(CarbonCommonConstants.CARBON_MAX_EXECUTOR_LRU_CACHE_SIZE) != null
-
+  private val operationContext: OperationContext = new OperationContext
```
Comment (kunal642) [inline integration/spark2/src/main/scala/org/apache/carbondata/indexserver/IndexServer.scala]: This operation context would be common for all the calls to index server. I think its safe to make this as a local variable wherever needed

**Comment by kumarvishal09** [comment]
Please add inline comments

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/606/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/607/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/614/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/610/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/615/

**Comment by kunal642** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/622/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/627/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/627/

**Comment by kumarvishal09** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/698/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/702/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/702/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/736/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/744/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/743/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/740/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/748/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/747/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/741/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/749/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/748/

**Comment by kumarvishal09** [comment]
LGTM

**Comment by akashrn5** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/808/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/815/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/812/

**Comment by akashrn5** [comment]
LGTM

**Pull request closed without merging**